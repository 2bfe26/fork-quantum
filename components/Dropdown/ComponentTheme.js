import { darken } from 'polished';

const ComponentColors = {
  DEFAULT: '#cce5ff',
  SUCCESS: '#d4edda',
  WARNING: '#fff3cd',
  ERROR: '#f8d7da',
  INFO: '#d1ecf1',
};

const ComponentTheme = {
  default: {
    backgroundColor: ComponentColors.DEFAULT,
    borderColor: darken(0.05, ComponentColors.DEFAULT),
    color: darken(0.5, ComponentColors.DEFAULT),
  },
  success: {
    backgroundColor: ComponentColors.SUCCESS,
    borderColor: darken(0.05, ComponentColors.SUCCESS),
    color: darken(0.5, ComponentColors.SUCCESS),
  },
  warning: {
    backgroundColor: ComponentColors.WARNING,
    borderColor: darken(0.05, ComponentColors.WARNING),
    color: darken(0.5, ComponentColors.WARNING),
  },
  error: {
    backgroundColor: ComponentColors.ERROR,
    borderColor: darken(0.05, ComponentColors.ERROR),
    color: darken(0.5, ComponentColors.ERROR),
  },
  info: {
    backgroundColor: ComponentColors.INFO,
    borderColor: darken(0.05, ComponentColors.INFO),
    color: darken(0.5, ComponentColors.INFO),
  },
};

export default ComponentTheme;
