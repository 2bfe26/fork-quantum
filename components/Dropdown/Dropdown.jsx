import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import Downshift from 'downshift';
import Icon from '../Icon/Icon';
import Colors from '../Colors';
import { FieldGroup, Label, ErrorMessage } from '../shared';

const ITEM_HEIGHT = '44px';

const DropLabel = styled(Label)`
  margin-bottom: 8px;
  padding-left: 13.5px;
`;

DropLabel.displayName = 'DropLabel';

const DropButton = styled.button`
  align-items: center;
  background-color: ${Colors.WHITE};
  border-radius: 4px;
  border: 1.5px solid ${Colors.BLACK['400']};
  color: ${Colors.BLACK['400']};
  cursor: pointer;
  display: flex;
  font-size: initial;
  height: ${ITEM_HEIGHT};
  justify-content: space-between;
  letter-spacing: 0.2px;
  padding: 10px 12px;
  transition: all 0.2s ease-in-out;
  width: 100%;

  :hover,
  :focus {
    border-color: ${Colors.BLUE['500']};
    box-shadow: 0 2px 6px 0 ${Colors.BLUE['50']};
  }

  ${({ error }) =>
    error &&
    `
    border-color: ${Colors.ERROR['500']};

    :hover, :focus {
      box-shadow: 0 2px 6px 0 ${Colors.ERROR['500']};
      border-color: ${Colors.ERROR['500']};
    }
  `};

  &[disabled] {
    background-color: ${Colors.BLACK['100']};
    color: ${Colors.BLACK['400']};
    cursor: not-allowed;
    border-color: ${Colors.BLACK['400']};
    box-shadow: none;
  }

  ${({ text }) => !text && 'flex-direction: row-reverse;'};
  ${({ selectedItem }) => selectedItem && `color: ${Colors.BLACK[700]}`};
`;

const ArrowDown = styled(Icon).attrs({
  name: 'keyboard_arrow_down',
})`
  color: ${Colors.BLACK['700']};
  font-size: 1.5em;
  pointer-events: none;
`;

const DropList = styled.ul`
  background-color: ${Colors.WHITE};
  box-shadow: 0 2px 6px 0 ${Colors.SHADOW[40]};
  list-style: none;
  margin-top: 4px;
  max-height: calc(${ITEM_HEIGHT} * 7);
  overflow: auto;
  padding: 0;
  position: absolute;
  width: 100%;
  z-index: 9999;
`;

const CheckIcon = styled(Icon).attrs({
  name: 'check',
})`
  color: ${Colors.BLUE['500']};
`;

const DropItem = styled.li`
  background-color: ${Colors.WHITE};
  border-radius: 4px;
  border: solid 1.5px ${Colors.BLACK[100]};
  border-bottom-width: 0;
  cursor: pointer;
  height: ${ITEM_HEIGHT};
  padding: 10px 15px;
  box-sizing: border-box;

  &[aria-selected='true'] {
    background-color: ${Colors.BLUE[200]};
  }

  ${({ isSelected }) =>
    isSelected &&
    `
    display: flex;
    justify-content: space-between;
  `}
`;

const DropError = styled(ErrorMessage)`
  margin-top: 8px;
  padding-left: 13.5px;
`;

const DropContainer = styled.div`
  position: relative;
`;

const RequiredMark = styled.em`
  color: ${Colors.ERROR['500']};
`;

const _getValue = item => (item ? item.value || item.label || item : '');
const _getLabel = item => (item ? item.label || item.value || item : '');
const _isEqual = (selected, item) => _getValue(selected) === _getValue(item);

const Dropdown = ({
  label,
  error,
  required,
  disabled,
  items,
  placeholder,
  selectedItem,
  onChange,
  ...rest
}) => {
  const _buttonLabel = selectedItem ? _getLabel(selectedItem) : placeholder;
  const _buttonRef = React.createRef();

  const _reducer = ({ selectedItem: selected }, changes) => {
    if (changes.isOpen !== undefined && changes.isOpen) {
      return {
        ...changes,
        highlightedIndex: items.map(_getValue).indexOf(_getValue(selected)),
      };
    }

    return changes;
  };

  return (
    <FieldGroup>
      <Downshift
        {...rest}
        selectedItem={selectedItem}
        onChange={onChange}
        itemToString={_getValue}
        stateReducer={_reducer}
      >
        {({
          getRootProps,
          getLabelProps,
          getInputProps,
          getToggleButtonProps,
          getItemProps,
          isOpen,
        }) => (
          <DropContainer {...getRootProps()}>
            {label && (
              <DropLabel
                {...getLabelProps()}
                onClick={() => _buttonRef.current.focus()}
              >
                {label}
                {required && <RequiredMark>*</RequiredMark>}
              </DropLabel>
            )}
            <input type="hidden" {...getInputProps()} />
            <DropButton
              {...getToggleButtonProps()}
              ref={_buttonRef}
              isOpen={isOpen}
              disabled={disabled}
              error={error}
              text={_buttonLabel}
              selectedItem={selectedItem}
            >
              {_buttonLabel}
              <ArrowDown />
            </DropButton>
            {isOpen && (
              <DropList>
                {items.map(item => (
                  <DropItem
                    {...getItemProps({
                      item,
                      isSelected: _isEqual(selectedItem, item),
                      key: _getValue(item),
                    })}
                  >
                    {_getLabel(item)}
                    {_isEqual(selectedItem, item) && <CheckIcon />}
                  </DropItem>
                ))}
              </DropList>
            )}
          </DropContainer>
        )}
      </Downshift>

      {error && <DropError>{error}</DropError>}
    </FieldGroup>
  );
};

Dropdown.defaultProps = {
  disabled: false,
  error: '',
  items: [],
  label: '',
  onChange: () => {},
  placeholder: 'Select an option',
  required: false,
  selectedItem: '',
};

const itemPropType = PropTypes.oneOfType([
  PropTypes.string,
  PropTypes.shape({
    value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    label: PropTypes.string,
  }),
]);

Dropdown.propTypes = {
  disabled: PropTypes.bool,
  error: PropTypes.string,
  items: PropTypes.arrayOf(itemPropType),
  label: PropTypes.string,
  onChange: PropTypes.func,
  placeholder: PropTypes.string,
  required: PropTypes.bool,
  selectedItem: itemPropType,
};

export default Dropdown;
