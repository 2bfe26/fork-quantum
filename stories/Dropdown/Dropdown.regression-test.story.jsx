import React from 'react';
import { storiesOf } from '@storybook/react';

import {
  Simple,
  AutoComplete,
  RequiredMark,
  WithError,
  Disabled,
  WithImages,
} from './examples';

storiesOf('Dropdown', module)
  .add('default', () => <Simple />)
  .add('autocomplete', () => <AutoComplete />)
  .add('required', () => <RequiredMark />)
  .add('error', () => <WithError />)
  .add('disabled', () => <Disabled />)
  .add('with images', () => <WithImages />);
