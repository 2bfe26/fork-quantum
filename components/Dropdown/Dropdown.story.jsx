import React from 'react';
import { storiesOf } from '@storybook/react';
import Heading from '../../.storybook/components/Heading';
import AutoPropsApi from '../../.storybook/components/AutoPropsApi';
import ComponentPanel from '../../.storybook/components/ComponentPanel';
import { TabbedView, Tab } from '../../.storybook/components/TabbedView';
import Dropdown from './Dropdown';

storiesOf('5. Selection', module)
  .add('Dropdown', () => (
    <Heading atom title="Dropdown">
      <TabbedView>
        <Tab title="Usage">
          <ComponentPanel component={<Dropdown />} importModules="Dropdown" />
        </Tab>

        <Tab title="API">
          <AutoPropsApi component={Dropdown} />
        </Tab>
      </TabbedView>
    </Heading>
  ));
