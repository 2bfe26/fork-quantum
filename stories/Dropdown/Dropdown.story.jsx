import React from 'react';
import PropTypes from 'prop-types';
import { storiesOf } from '@storybook/react';
import {
  AutoPropsApi,
  Heading,
  SimpleHighlight,
  StoryContainer,
  Tab,
  TabbedView,
  Title,
} from '@catho/quantum-storybook-ui';

import Dropdown from '../../components/Dropdown';
import { Row, Col } from '../../components/Grid';
import List from '../../components/List';
import {
  Simple,
  CustomLabel,
  RequiredMark,
  WithError,
  Disabled,
  Controlled,
  AutoComplete,
} from './examples';

const DropdownExample = ({ component: Component }) => (
  <Row>
    <Col xsmall={2} small={4} medium={6}>
      <SimpleHighlight>{Component.CODE}</SimpleHighlight>
    </Col>
    <Col xsmall={2} small={4} medium={6}>
      <Component />
    </Col>
  </Row>
);

DropdownExample.propTypes = {
  component: PropTypes.func.isRequired,
};

const importCode = "import { Dropdown } from '@catho/quantum';";

storiesOf('Forms', module).add('Dropdown', () => (
  <>
    <Heading name="Dropdown">
      A dropdown receives user input, allowing it to choose one of several
      options.
    </Heading>

    <TabbedView>
      <Tab title="Usage">
        <StoryContainer>
          <Title as="h2">Importing</Title>
          <SimpleHighlight>{importCode}</SimpleHighlight>
          <div>
            <p>
              Its options are described by {'"'}items{'"'} prop, which can be
              either array of string or objects, within the following structure:
            </p>
            <List>
              <List.Item>
                <strong>value</strong>: Computed to dropdown value itself, if
                the corresponding item is selected
              </List.Item>
              <List.Item>
                <strong>label</strong>: What is showed to the user in the
                dropdown interface
              </List.Item>
            </List>
            <p>
              If a list of strings is provided, they will be used both for items
              values and labels.
            </p>
          </div>
          <DropdownExample component={Simple} />
          <DropdownExample component={CustomLabel} />
          <DropdownExample component={AutoComplete} />
          <DropdownExample component={RequiredMark} />
          <DropdownExample component={WithError} />
          <DropdownExample component={Disabled} />
          <DropdownExample component={Controlled} />
        </StoryContainer>
      </Tab>
      <Tab title="API">
        <AutoPropsApi component={Dropdown} />
      </Tab>
    </TabbedView>
  </>
));
