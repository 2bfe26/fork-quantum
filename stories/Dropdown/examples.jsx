import React from 'react';
import PropTypes from 'prop-types';
import Dropdown from '../../components/Dropdown';

class Controlled extends React.Component {
  static CODE = `class ControlledDropdown extends React.Component {
  constructor(props) {
    super(props);

    const { selectedItem = '' } = props;

    this.state = { selectedItem };
  }

  render() {
    const { selectedItem } = this.state;

    return (
      <Dropdown
        label="Controlled"
        placeholder="Select a district"
        selectedItem={selectedItem}
        onChange={(item) => this.setState({ selectedItem: item })}
        items={[
          { label: 'Amazonas', value: 'AM' },
          { label: 'Bahia', value: 'BA' },
          { label: 'Ceará', value: 'CE' },
          { label: 'Goiás', value: 'GO' },
          { label: 'Minas Gerais', value: 'MG' },
          { label: 'Paraná', value: 'PR' },
          { label: 'Pernambuco', value: 'PE' },
          { label: 'Rio de Janeiro', value: 'RJ' },
          { label: 'Rio Grande do Sul', value: 'RS' },
          { label: 'Santa Catarina', value: 'SC' },
          { label: 'São Paulo', value: 'SP' },
        ]}
      />
    )
  }
}`;

  constructor(props) {
    super(props);

    const { selectedItem = '' } = props;

    this.state = { selectedItem };
  }

  render() {
    const { selectedItem } = this.state;

    return (
      <Dropdown
        label="Controlled"
        placeholder="Select a district"
        selectedItem={selectedItem}
        onChange={item => this.setState({ selectedItem: item })}
        items={[
          { label: 'Amazonas', value: 'AM' },
          { label: 'Bahia', value: 'BA' },
          { label: 'Ceará', value: 'CE' },
          { label: 'Goiás', value: 'GO' },
          { label: 'Minas Gerais', value: 'MG' },
          { label: 'Paraná', value: 'PR' },
          { label: 'Pernambuco', value: 'PE' },
          { label: 'Rio de Janeiro', value: 'RJ' },
          { label: 'Rio Grande do Sul', value: 'RS' },
          { label: 'Santa Catarina', value: 'SC' },
          { label: 'São Paulo', value: 'SP' },
        ]}
      />
    );
  }
}

Controlled.defaultProps = {
  selectedItem: '',
};

Controlled.propTypes = {
  selectedItem: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
};

const Simple = () => (
  <Dropdown items={['Lemon', 'Banana', 'Strawberry', 'Orange', 'Avocado']} />
);
Simple.CODE = `<Dropdown
  items={['Lemon', 'Banana', 'Strawberry', 'Orange', 'Avocado']}
/>`;

const CustomLabel = () => (
  <Dropdown
    label="With a custom label"
    items={['Lemon', 'Banana', 'Strawberry', 'Orange', 'Avocado']}
  />
);
CustomLabel.CODE = `<Dropdown
  label="With a custom label"
  items={['Lemon', 'Banana', 'Strawberry', 'Orange', 'Avocado']}
/>`;

const RequiredMark = () => (
  <Dropdown
    required
    label="With a required mark"
    items={['Lemon', 'Banana', 'Strawberry', 'Orange', 'Avocado']}
  />
);
RequiredMark.CODE = `<Dropdown
  required
  label="With a required mark"
  items={['Lemon', 'Banana', 'Strawberry', 'Orange', 'Avocado']}
/>`;

const WithError = () => (
  <Dropdown
    error="Don't worry, this is just a message error example ;)"
    label="With an error"
    items={['Lemon', 'Banana', 'Strawberry', 'Orange', 'Avocado']}
  />
);
WithError.CODE = `<Dropdown
  error="Don't worry, this is just a message error example ;)"
  label="With an error"
  items={['Lemon', 'Banana', 'Strawberry', 'Orange', 'Avocado']}
/>`;

const Disabled = () => (
  <Dropdown
    disabled
    placeholder="What fruit do you prefer?"
    label="Disabled, with a custom placeholder"
    items={['Lemon', 'Banana', 'Strawberry', 'Orange', 'Avocado']}
  />
);
Disabled.CODE = `<Dropdown
  disabled
  placeholder="What fruit do you prefer?"
  label="Disabled, with a custom placeholder"
  items={['Lemon', 'Banana', 'Strawberry', 'Orange', 'Avocado']}
/>`;

const AutoComplete = () => (
  <Dropdown
    autocomplete
    label="Autocomplete"
    placeholder="Start typing..."
    items={['Lemon', 'Banana', 'Strawberry', 'Orange', 'Avocado']}
  />
);
AutoComplete.CODE = `<Dropdown
  autocomplete
  label="Autocomplete"
  placeholder="Start typing..."
  items={['Lemon', 'Banana', 'Strawberry', 'Orange', 'Avocado']}
/>`;

const WithImages = () => (
  <Dropdown
    placeholder="What credit card do you prefer?"
    label="choose a credit card"
    items={[
      {
        label: 'Master card',
        value: 'master_0',
        img: 'https://dummyimage.com/24x24',
        alt: 'descrição da imagem',
      },
      {
        label: 'Master card',
        value: 'master_1',
        img: 'https://ca.slack-edge.com/T04S23U24-UL9UMHYNM-760c576f6716-72',
        alt: 'descrição da imagem',
      },
      {
        label: 'Master card',
        value: 'master_2',
        img: 'https://ca.slack-edge.com/T04S23U24-UL9UMHYNM-760c576f6716-72',
        alt: 'descrição da imagem',
      },
    ]}
  />
);
WithImages.CODE = `
<Dropdown
placeholder="What credit card do you prefer?"
label="choose a credit card"
items={[
  { label: 'Master card', value: 'master', img: '' },
  { label: 'Master card', value: 'master', img: '' },
  { label: 'Master card', value: 'master', img: '' },
]}
/>
`;

export {
  Simple,
  CustomLabel,
  RequiredMark,
  WithError,
  Disabled,
  Controlled,
  AutoComplete,
  WithImages,
};
