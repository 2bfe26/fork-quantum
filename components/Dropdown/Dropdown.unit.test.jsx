import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import Dropdown from './Dropdown';

describe('Input component ', () => {
  it('should match the snapshot', () => {
    const withItems = (
      <Dropdown
        items={[
          { value: 'fooKey', label: 'foo' },
          { value: 'barKey', label: 'bar' },
          { value: 'bazKey', label: 'baz' },
        ]}
      />
    );

    expect(renderer.create(<Dropdown />).toJSON()).toMatchSnapshot();
    expect(
      renderer.create(<Dropdown placeholder="Dropdown placeholder" />).toJSON(),
    ).toMatchSnapshot();
    expect(
      renderer.create(<Dropdown placeholder="" />).toJSON(),
    ).toMatchSnapshot();
    expect(
      renderer.create(<Dropdown error="Error message" />).toJSON(),
    ).toMatchSnapshot();
    expect(renderer.create(<Dropdown disabled />).toJSON()).toMatchSnapshot();
    expect(
      renderer.create(<Dropdown label="Dropdown label" />).toJSON(),
    ).toMatchSnapshot();
    expect(
      renderer.create(<Dropdown required label="Dropdown label" />).toJSON(),
    ).toMatchSnapshot();
    expect(renderer.create(withItems).toJSON()).toMatchSnapshot();
  });
});

describe('with an "onChange" callback set', () => {
  const mockFn = jest.fn();

  const items = [
    {
      value: 'fooKey',
      label: 'foo',
    },
    {
      value: 'barKey',
      label: 'bar',
    },
  ];

  const dropdown = <Dropdown onChange={mockFn} id="dropdown" items={items} />;

  const wrapper = shallow(dropdown);

  it('should init value with an empty label', () => {
    expect(wrapper.state('selectedItem')).toEqual(null);
  });

  it('should call the callback and set a new value', () => {
    const [selectedItem] = items;

    wrapper.find('#dropdown').simulate('change', selectedItem);

    expect(wrapper.state('selectedItem')).toEqual(selectedItem);
    expect(mockFn).toHaveBeenCalledTimes(1);
    expect(mockFn).toBeCalledWith(null, { selectedItem });
  });
});
