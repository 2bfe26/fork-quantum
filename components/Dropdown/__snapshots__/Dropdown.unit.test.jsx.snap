// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input component  should match the snapshot 1`] = `
<div
  className="FieldGroup-iFQzIc dToppz"
>
  
  <select
    className="Dropdown__Select-kLQAl TvOMx"
    id="dropdown"
    name="Dropdown"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    value=""
  >
    <option
      disabled={true}
      value=""
    >
      Select...
    </option>
  </select>
  <span
    aria-hidden="true"
    className="material-icons MuiIcon-root-1 Dropdown__ArrowDown-laZZPE kfWEYF"
    style={
      Object {
        "color": "#263238",
      }
    }
  >
    keyboard_arrow_down
  </span>
  
</div>
`;

exports[`Input component  with a label should match the snapshot 1`] = `
<div
  className="FieldGroup-iFQzIc dToppz"
>
  <label
    className="Label-bxMjsr iZhmvb"
    htmlFor="dropdown"
  >
     
    Text label
     
  </label>
  <select
    className="Dropdown__Select-kLQAl TvOMx"
    id="dropdown"
    name="Dropdown"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    value=""
  >
    <option
      disabled={true}
      value=""
    >
      Select...
    </option>
  </select>
  <span
    aria-hidden="true"
    className="material-icons MuiIcon-root-1 Dropdown__ArrowDown-laZZPE kfWEYF"
    style={
      Object {
        "color": "#263238",
      }
    }
  >
    keyboard_arrow_down
  </span>
  
</div>
`;

exports[`Input component  with an error message should match the snapshot 1`] = `
<div
  className="FieldGroup-iFQzIc dToppz"
>
  
  <select
    className="Dropdown__Select-kLQAl TvOMx"
    id="dropdown"
    name="Dropdown"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    value=""
  >
    <option
      disabled={true}
      value=""
    >
      Select...
    </option>
  </select>
  <span
    aria-hidden="true"
    className="material-icons MuiIcon-root-1 Dropdown__ArrowDown-laZZPE kfWEYF"
    style={
      Object {
        "color": "#263238",
      }
    }
  >
    keyboard_arrow_down
  </span>
  <div
    className="ErrorMessage-eJhZpN bzMSZo"
  >
    Error message
  </div>
</div>
`;

exports[`Input component  with items should match the snapshot 1`] = `
<div
  className="FieldGroup-iFQzIc dToppz"
>
  
  <select
    className="Dropdown__Select-kLQAl TvOMx"
    id="dropdown"
    name="Dropdown"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    value=""
  >
    <option
      disabled={true}
      value=""
    >
      Select...
    </option>
    <option
      value="fooKey"
    >
      foo
    </option>
    <option
      value="barKey"
    >
      bar
    </option>
    <option
      value="bazKey"
    >
      baz
    </option>
  </select>
  <span
    aria-hidden="true"
    className="material-icons MuiIcon-root-1 Dropdown__ArrowDown-laZZPE kfWEYF"
    style={
      Object {
        "color": "#263238",
      }
    }
  >
    keyboard_arrow_down
  </span>
  
</div>
`;
